src/fabric/test/fabric2_index_tests.erl

```diff
@@ -0,0 +1,302 @@
+    catch fabric2_db:delete(fabric2_db:name(Db1), []),
+    catch fabric2_db:delete(fabric2_db:name(Db2), []),
+
+    test_util:stop_couch(Ctx),
+    application:set_env(fabric, indices, Indices),
+
+    meck:unload().
+
+
+register_index_works(_) ->
+    reset_callbacks(),
+
+    Mod1 = fabric2_test_callback1,
+    fabric2_index:register_index(Mod1),
+    Indices1 = application:get_env(fabric, indices, []),
+    ?assertEqual([Mod1], Indices1),
+
+    Mod2 = fabric2_test_callback2,
+    fabric2_index:register_index(Mod2),
+    Indices2 = application:get_env(fabric, indices, []),
+    ?assertEqual(lists:sort([Mod1, Mod2]), lists:sort(Indices2)).
+
+
+single_update(#{db1 := Db}) ->
+    reset_callbacks(),
+
+    Mod = fabric2_test_callback3,
+    setup_callback(Mod),
+    create_doc(Db),
+
+    meck:wait(Mod, build_indices, 2, 2000),
+    ?assertEqual(1, meck:num_calls(Mod, build_indices, 2)).
+
+
+multiple_updates(#{db1 := Db}) ->
+    reset_callbacks(),
+
+    Mod = fabric2_test_callback4,
+    setup_callback(Mod),
+    create_docs(Db, 10),
+
+    % should be called at least once
+    meck:wait(Mod, build_indices, 2, 2000),
+
+    % Maybe called another time or two at most
+    timer:sleep(500),
+    ?assert(meck:num_calls(Mod, build_indices, 2) =< 3).
+
+
+skip_db_if_no_ddocs(#{db2 := Db}) ->
+    reset_callbacks(),
+
+    Mod = fabric2_test_callback5,
+    setup_callback(Mod),
+    create_doc(Db),
+
+    timer:sleep(500),
+    ?assertEqual(0, meck:num_calls(Mod, build_indices, 2)).
+
+
+ignore_deleted_dbs(#{}) ->
+    reset_callbacks(),
+
+    Mod = fabric2_test_callback6,
+    setup_callback(Mod),
+    lists:foreach(fun(_) ->
+        RandomDbName = fabric2_util:uuid(),
+        fabric2_index:db_updated(RandomDbName)
+    end, lists:seq(1, 10000)),
+
+    test_util:wait(fun() ->
+        case table_sizes() =:= 0 of
+            true -> ok;
+            false -> wait
+        end
+    end, 5000).
+
+
+check_gen_server_messages(#{}) ->
+    CallExpect = {stop, {bad_call, foo}, {bad_call, foo}, baz},
+    CastExpect = {stop, {bad_cast, foo}, bar},
+    InfoExpect = {stop, {bad_info, foo}, bar},
+    ?assertEqual(CallExpect, fabric2_index:handle_call(foo, bar, baz)),
+    ?assertEqual(CastExpect, fabric2_index:handle_cast(foo, bar)),
+    ?assertEqual(InfoExpect, fabric2_index:handle_info(foo, bar)),
+    ?assertEqual(ok, fabric2_index:terminate(shutdown, nil)),
+    ?assertEqual({ok, nil}, fabric2_index:code_change(v0, nil, extra)).
+
+
+updater_processes_start(#{}) ->
+    Pid = whereis(fabric2_index),
+    ?assert(is_process_alive(Pid)),
+    lists:map(fun(N) ->
+        ?assertEqual(tid(N), ets:info(tid(N), name))
+    end, lists:seq(0, ?SHARDS - 1)).
+
+
+updater_processes_stop(#{}) ->
+    Refs = lists:map(fun(N) ->
+        Pid = ets:info(tid(N), owner),
+        ?assert(is_process_alive(Pid)),
+        monitor(process, Pid)
+    end, lists:seq(0, ?SHARDS - 1)),
+
+    application:stop(fabric),
```

Comment:
You don't have to restart fabric after this?

Reply:
Oh, subtle, you have it as a foreach on the second group. Might make a note of that here since its a subtle thing to miss and up way above.

Reply:
Yeah it's because a few tests are running in a foreach-type fixture since they kill stop the app and do other nasty things to the application and it was easier ensure they don't affect other tests.

I'll add a comment in the test why we are not restarting the app there.